File: modules/core/src/main/scala/com/github/chengpohi/context/EDQLDefinition.scala
package com.github.chengpohi.context

import com.github.chengpohi.edql.parser.json.JsonCollection
import org.apache.http.util.EntityUtils
import org.elasticsearch.client.{Request, RequestOptions, ResponseException}

import scala.concurrent.Future

/**
 * eql
 * Created by chengpohi on 6/28/16.
 */
trait EDQLDefinition extends ElasticBase with EDQLExecutor with FutureOps {
  val KIBANA_PROXY_METHOD: String = "KIBANA_PROXY_METHOD"
  val KIBANA_PATH_PREFIX: String = "KIBANA_PATH_PREFIX"

  case class GetActionDefinition(path: String, action: Option[JsonCollection.Val])
    extends Definition[String] {
    override def execute: Future[String] = {
      val as = action.filter(_.isInstanceOf[JsonCollection.Obj]).map(_.asInstanceOf[JsonCollection.Obj])
      val request = new Request(if (kibanaProxy) "POST" else "GET", path);

      request.setOptions(RequestOptions.DEFAULT.toBuilder
        .addHeader(KIBANA_PROXY_METHOD, "GET")
        .addHeader(KIBANA_PATH_PREFIX, pathPrefix)
      );

      as match {
        case None =>
        case Some(a) =>
          request.setJsonEntity(a.toJson + System.lineSeparator())
      }
      Future {
        try {
          val entity = restClient.performRequest(request).getEntity
          EntityUtils.toString(entity)
        } catch {
          case ex: ResponseException =>
            val responseEntityStr = EntityUtils.toString(ex.getResponse.getEntity)
            if (responseEntityStr.isBlank) {
              ex.getMessage
            } else {
              responseEntityStr
            }
        }
      }
    }

    override def json: String = execute.await.json
  }

  case class HeadActionDefinition(path: String, action: Option[String])
    extends Definition[String] {
    override def execute: Future[String] = {
      val request = new Request(if (kibanaProxy) "POST" else "HEAD", path);
      request.setOptions(RequestOptions.DEFAULT.toBuilder
        .addHeader(KIBANA_PROXY_METHOD, "HEAD")
        .addHeader(KIBANA_PATH_PREFIX, pathPrefix)
      );
      request.setJsonEntity(action.orNull)
      Future {
        try {
          val entity = restClient.performRequest(request).getEntity
          EntityUtils.toString(entity)
        } catch {
          case ex: ResponseException =>
            val responseEntityStr = EntityUtils.toString(ex.getResponse.getEntity)
            if (responseEntityStr.isBlank) {
              ex.getMessage
            } else {
              responseEntityStr
            }
        }
      }
    }

    override def json: String = execute.await.json
  }


  case class PostActionDefinition(path: String, action: Seq[JsonCollection.Val])
    extends Definition[String] {
    override def execute: Future[String] = {
      val request = new Request("POST", path);

      request.setOptions(RequestOptions.DEFAULT.toBuilder
        .addHeader(KIBANA_PROXY_METHOD, "POST")
        .addHeader(KIBANA_PATH_PREFIX, pathPrefix)
      );

      val as = action.filter(_.isInstanceOf[JsonCollection.Obj]).map(_.asInstanceOf[JsonCollection.Obj])
      as match {
        case Seq() =>
        case a =>
          if (a.size > 1) {
            request.setJsonEntity(
              a.map(i => {
                trimQueryClause(i).toJson
              }).mkString(System.lineSeparator()) + System.lineSeparator())
          } else if (a.size == 1) {
            val h = trimQueryClause(a.head)
            request.setJsonEntity(h.toJson + System.lineSeparator())
          }
      }
      Future {
        try {
          val entity = restClient.performRequest(request).getEntity
          EntityUtils.toString(entity)
        }
        catch {
          case ex: ResponseException =>
            val responseEntityStr = EntityUtils.toString(ex.getResponse.getEntity)
            if (responseEntityStr.isBlank) {
              ex.getMessage
            } else {
              responseEntityStr
            }
        }
      }
    }

    private def trimQueryClause(i: JsonCollection.Obj): JsonCollection.Obj = {
      val emptyQueryClause = i.get("query").flatMap(_ match {
        case value: JsonCollection.Var =>
          value.realValue
        case a => Some(a)
      }).filter(i => i.isInstanceOf[JsonCollection.Obj]).exists(i => {
        i.asInstanceOf[JsonCollection.Obj].value.isEmpty
      })

      emptyQueryClause match {
        case true => i.remove("query")
        case false => i
      }
    }

    override def json: String = execute.await.json
  }

  case class PutActionDefinition(path: String, action: Option[String])
    extends Definition[String] {
    override def execute: Future[String] = {
      val request = new Request(if (kibanaProxy) "POST" else "PUT", path);
      request.setOptions(RequestOptions.DEFAULT.toBuilder
        .addHeader(KIBANA_PROXY_METHOD, "PUT")
        .addHeader(KIBANA_PATH_PREFIX, pathPrefix)
      );
      request.setJsonEntity(action.orNull)
      Future {
        try {
          val entity = restClient.performRequest(request).getEntity
          EntityUtils.toString(entity)
        } catch {
          case ex: ResponseException =>
            val responseEntityStr = EntityUtils.toString(ex.getResponse.getEntity)
            if (responseEntityStr.isBlank) {
              ex.getMessage
            } else {
              responseEntityStr
            }
        }
      }
    }

    override def json: String = execute.await.json
  }

  case class DeleteActionDefinition(path: String, action: Option[String])
    extends Definition[String] {
    override def execute: Future[String] = {
      val request = new Request(if (kibanaProxy) "POST" else "DELETE", path);
      request.setOptions(RequestOptions.DEFAULT.toBuilder
        .addHeader(KIBANA_PROXY_METHOD, "DELETE")
        .addHeader(KIBANA_PATH_PREFIX, pathPrefix)
      );
      request.setJsonEntity(action.orNull)
      Future {
        try {
          val entity = restClient.performRequest(request).getEntity
          EntityUtils.toString(entity)
        } catch {
          case ex: ResponseException =>
            val responseEntityStr = EntityUtils.toString(ex.getResponse.getEntity)
            if (responseEntityStr.isBlank) {
              ex.getMessage
            } else {
              responseEntityStr
            }
        }
      }
    }

    override def json: String = execute.await.json
  }


  case class ParserErrorDefinition(parameters: Seq[JsonCollection.Val])
    extends Definition[Map[String, AnyRef]] {

    override def execute: Future[Map[String, AnyRef]] = {
      val res = List("error_msg", "caused_by")
        .zip(parameters.take(2).map(_.extract[String]).toList)
        .toMap
      Future.successful(res)
    }

    override def json: String = execute.await.json
  }

}

File: src/main/java/com/github/chengpohi/context/EDQLDefinition.kt
package com.github.chengpohi.context

import com.github.chengpohi.edql.parser.json.JsonCollection
import com.github.chengpohi.script.ScriptContext
import org.apache.http.util.EntityUtils
import org.elasticsearch.client.Request
import org.elasticsearch.client.RequestOptions
import org.elasticsearch.client.ResponseException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext

object EDQLDefinition {
    const val KIBANA_PROXY_METHOD: String = "KIBANA_PROXY_METHOD"
    const val KIBANA_PATH_PREFIX: String = "KIBANA_PATH_PREFIX"
    const val READONLY_MODE: String = "readonly mode"

    data class GetActionDefinition(
        val path: String,
        val action: JsonCollection.Val?
    ) : Definition<String> {
        override suspend fun execute(context: ScriptContext): String = withContext(Dispatchers.IO) {
            val asObj = action?.let { if (it is JsonCollection.Obj) it else null }
            val request = Request(if (context.kibanaProxy) "POST" else "GET", path)

            val builder = RequestOptions.DEFAULT.toBuilder()
            builder.addHeader("Content-Type", "application/json")
            if (context.kibanaProxy) {
                builder.addHeader(KIBANA_PROXY_METHOD, "GET")
                    .addHeader(KIBANA_PATH_PREFIX, context.pathPrefix)
            }
            request.options = builder.build()

            asObj?.let {
                request.setJsonEntity(it.toJson() + System.lineSeparator())
            }

            try {
                val entity = context.client.restClient.performRequest(request).entity
                EntityUtils.toString(entity)
            } catch (ex: ResponseException) {
                val responseEntityStr = EntityUtils.toString(ex.response.entity)
                if (responseEntityStr.isBlank()) {
                    ex.message ?: ""
                } else {
                    responseEntityStr
                }
            }
        }

        override fun json(context: ScriptContext): String {
            return runBlocking { execute(context) }
        }
    }

    data class HeadActionDefinition(
        val path: String,
        val action: String?
    ) : Definition<String> {
        override suspend fun execute(context: ScriptContext): String = withContext(Dispatchers.IO) {
            if (readOnlyPath(path, context.readOnly)) {
                throw RuntimeException(READONLY_MODE)
            }

            val request = Request(if (context.kibanaProxy) "POST" else "HEAD", path)
            if (context.kibanaProxy) {
                request.options = RequestOptions.DEFAULT.toBuilder()
                    .addHeader(KIBANA_PROXY_METHOD, "HEAD")
                    .addHeader(KIBANA_PATH_PREFIX, context.pathPrefix)
                    .build()
            }
            request.setJsonEntity(action)

            try {
                val entity = context.client.restClient.performRequest(request).entity
                EntityUtils.toString(entity)
            } catch (ex: ResponseException) {
                val responseEntityStr = EntityUtils.toString(ex.response.entity)
                if (responseEntityStr.isBlank()) {
                    ex.message ?: ""
                } else {
                    responseEntityStr
                }
            }
        }

        override fun json(context: ScriptContext): String {
            return runBlocking { execute(context) }
        }
    }

    data class PostActionDefinition(
        val path: String,
        val action: List<JsonCollection.Val>
    ) : Definition<String> {
        override suspend fun execute(context: ScriptContext): String = withContext(Dispatchers.IO) {
            if (readOnlyPath(path, context.readOnly)) {
                throw RuntimeException(READONLY_MODE)
            }

            val request = Request("POST", path)

            if (context.kibanaProxy) {
                request.options = RequestOptions.DEFAULT.toBuilder()
                    .addHeader(KIBANA_PROXY_METHOD, "POST")
                    .addHeader(KIBANA_PATH_PREFIX, context.pathPrefix)
                    .build()
            }

            val asObj = action.filterIsInstance<JsonCollection.Obj>()
            when {
                asObj.isEmpty() -> {
                    // Do nothing
                }
                asObj.size > 1 -> {
                    request.setJsonEntity(
                        asObj.joinToString(System.lineSeparator()) { trimQueryClause(it).toJson() } + System.lineSeparator()
                    )
                }
                else -> {
                    val h = trimQueryClause(asObj.first())
                    request.setJsonEntity(h.toJson() + System.lineSeparator())
                }
            }

            try {
                val entity = context.client.restClient.performRequest(request).entity
                EntityUtils.toString(entity)
            } catch (ex: ResponseException) {
                val responseEntityStr = EntityUtils.toString(ex.response.entity)
                if (responseEntityStr.isBlank()) {
                    ex.message ?: ""
                } else {
                    responseEntityStr
                }
            }
        }

        override fun json(context: ScriptContext): String {
            return runBlocking { execute(context) }
        }
    }

    data class PutActionDefinition(
        val path: String,
        val action: List<JsonCollection.Val>
    ) : Definition<String> {
        override suspend fun execute(context: ScriptContext): String = withContext(Dispatchers.IO) {
            if (context.readOnly) {
                throw RuntimeException(READONLY_MODE)
            }
            val request = Request(if (context.kibanaProxy) "POST" else "PUT", path)
            if (context.kibanaProxy) {
                request.options = RequestOptions.DEFAULT.toBuilder()
                    .addHeader(KIBANA_PROXY_METHOD, "PUT")
                    .addHeader(KIBANA_PATH_PREFIX, context.pathPrefix)
                    .build()
            }
            val asObj = action.filterIsInstance<JsonCollection.Obj>()
            when {
                asObj.isEmpty() -> {
                    // Do nothing
                }
                asObj.size > 1 -> {
                    request.setJsonEntity(
                        asObj.joinToString(System.lineSeparator()) { trimQueryClause(it).toJson() } + System.lineSeparator()
                    )
                }
                else -> {
                    val h = trimQueryClause(asObj.first())
                    request.setJsonEntity(h.toJson() + System.lineSeparator())
                }
            }

            try {
                val entity = context.client.restClient.performRequest(request).entity
                EntityUtils.toString(entity)
            } catch (ex: ResponseException) {
                val responseEntityStr = EntityUtils.toString(ex.response.entity)
                if (responseEntityStr.isBlank()) {
                    ex.message ?: ""
                } else {
                    responseEntityStr
                }
            }
        }

        override fun json(context: ScriptContext): String {
            return runBlocking { execute(context) }
        }
    }

    data class DeleteActionDefinition(
        val path: String,
        val action: String?
    ) : Definition<String> {
        override suspend fun execute(context: ScriptContext): String = withContext(Dispatchers.IO) {
            if (context.readOnly) {
                throw RuntimeException(READONLY_MODE)
            }
            val request = Request(if (context.kibanaProxy) "POST" else "DELETE", path)
            if (context.kibanaProxy) {
                request.options = RequestOptions.DEFAULT.toBuilder()
                    .addHeader(KIBANA_PROXY_METHOD, "DELETE")
                    .addHeader(KIBANA_PATH_PREFIX, context.pathPrefix)
                    .build()
            }
            request.setJsonEntity(action)

            try {
                val entity = context.client.restClient.performRequest(request).entity
                EntityUtils.toString(entity)
            } catch (ex: ResponseException) {
                val responseEntityStr = EntityUtils.toString(ex.response.entity)
                if (responseEntityStr.isBlank()) {
                    ex.message ?: ""
                } else {
                    responseEntityStr
                }
            }
        }

        override fun json(context: ScriptContext): String {
            return runBlocking { execute(context) }
        }
    }

    private fun readOnlyPath(path: String, readonly: Boolean): Boolean {
        if (!readonly) {
            return false
        }
        return path.contains("/_update") || path.contains("/_delete") || path.contains("/_bulk")
    }

    private fun trimQueryClause(i: JsonCollection.Obj): JsonCollection.Obj {
        val emptyQueryClause = i.get("query")?.let { value ->
            when (value) {
                is JsonCollection.Var -> value.realValue
                else -> value
            }
        }?.let { it is JsonCollection.Obj && it.value.isEmpty() } ?: false

        return if (emptyQueryClause) {
            i.remove("query")
        } else {
            i
        }
    }
} 